### ISSUE TITLE
Concurrent calls with em-synchrony don't return the correct results.

### ISSUE THREAD

**Issue by NeilBetham** [redis/redis-rb#499]
Hi,

I'm currently using redis with the em-synchrony driver inside an event machine driven application that i'm writing. I've been having an issue where two separate fibers are getting each others results swapped from requests they make to redis.

As a little background my application has a single instance of the redis client for all the calls inside the event machine reactor. The event machine reactor runs some periodic background tasks which all run in their own fibers as well as a small sinatra app running on thin inside the reactor loop. All these components make requests to redis and sometimes the requests happen concurrently.

Now this is where i'm having issues. In my application I periodically have commands running that I store the output to a redis list; it uses the `llen` command to add an index to the output. When a sinatra request comes in for the list data it will request the contents of that list to serve them to a client. However what usually happens is the callback that is writing the output gets the return of the `llen` command and sinatra gets the return of the `lrange` command.

Now, my main question, is this something that is inherent to the em-synchrony driver, and i'm using the client incorrectly or is there a problem with the driver that needs to be fixed?

TL;DR: concurrent calls to a single redis client in em-synchrony don't always return the correct results.

Thanks,
Neil

**Comment by NeilBetham** [comment]
I ended up switching clients to em-hiredis and used em-synchrony's sync call to get the same effect as the synchrony driver while getting the correct returns from requests.

**Comment by lashd** [comment]
I'm pretty sure I am seeing this problem also. GET commands are receiving the 'OK's of concurrent SET commands. This is then incorrectly unmarshalled and results in the following error:

`TypeError: incompatible marshal file format (can't be read)	format version 4.8 required; 79.75 given`

Connections are retrieved from a connection pool using the solution proposed in #185 but we still see the problem. Can anyone shed any light on whether there is still an issue here?

**Comment by plashchynski** [comment]
I have the same issue:
```
  incompatible marshal file format (can't be read)
        format version 4.8 required; 79.75 given
```
while using jbuilder json.cache! with redis cache store.

**Comment by NeilBetham** [comment]
So I was originally not using the connection pool, which was probably my issue. I now use 

```
REDIS = EventMachine::Synchrony::ConnectionPool.new(size: 10) do
  Redis.new driver: :synchrony
end
```

Once the pool is setup use I `REDIS` as normal and have had 0 issues so far.
